spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/DefaultRowWriter.scala

```diff
@@ -19,13 +19,12 @@ class DefaultRowWriter[T : TypeTag : ColumnMapper](
 
   override def readColumnValues(data: T, buffer: Array[Any]) = {
     val row = converter.convert(data)
-    for (i <- columnNames.indices)
+    for (i <- row.columnValues.indices)
```

Comment:
It shouldn't be changing the order, since these are the indices of the Seq and all of the actual data access is being done on the next line where members of the row and buffer are being accessed/modified.  In most cases, the number of table columns will be equal to the number of values in the row.  However, in the case that we're interested in (the underspecified case class case), the number of columns in the table will be greater than the number of values in the row, and so we want to use the smaller number.